airbyte-integrations/connectors/source-github/source_github/utils.py

```diff
@@ -24,3 +30,57 @@ def read_full_refresh(stream_instance: Stream):
         records = stream_instance.read_records(stream_slice=_slice, sync_mode=SyncMode.full_refresh)
         for record in records:
             yield record
+
+
+class MultipleTokenAuthenticatorWithRateLimiter(AbstractHeaderAuthenticator):
+    """
+    Each token in the cycle is checked against the rate limiter.
+    If a token exceeds the capacity limit, the system switches to another token.
+    If all tokens are exhausted, the system will enter a sleep state until
+    the first token becomes available again.
+    """
+
+    DURATION = 3600  # seconds
+
+    def __init__(self, tokens: List[str], requests_per_hour: int, auth_method: str = "Bearer", auth_header: str = "Authorization"):
+        self._auth_method = auth_method
+        self._auth_header = auth_header
+        now = time.time()
+        self._requests_per_hour = requests_per_hour
+        self._tokens = {t: SimpleNamespace(count=self._requests_per_hour, update_at=now) for t in tokens}
+        self._tokens_iter = cycle(self._tokens)
+
+    @property
+    def auth_header(self) -> str:
+        return self._auth_header
+
+    @property
+    def token(self) -> str:
+        while True:
+            token = next(self._tokens_iter)
+            if self._check_token(token):
+                return f"{self._auth_method} {token}"
```

Comment:
I'm not sure to understand how you perform:
> If a token exceeds the capacity limit, the system switches to another token.

From reading the code I've the impression that the execution is blocked sleeping when a token is exhausted and that you're not returning the next one that might not be exhausted.
I might be wrong. But I think your code would deserver additional docstring and testing 😄 .

Reply:
I have improved unit test, now test shows that we take another token if first one is exhausted
I have added docs for `def _sleep`